### ISSUE TITLE
cloudfodder relay list refresh issues

### ISSUE THREAD

**Issue by mikedilger** [mikedilger/gossip#694]
On master 625381969a22e777d9c64bb4dfa3ac265f82438b,

He has a new account following cloudfodder npub10npj3gydmv40m70ehemmal6vsdyfl7tewgvz043g54p0x23y0s8qzztl5h
but it doesn't have the relay list, and it never seems to refresh it.

He is using wss://nostr21.com/ as his discovery.

**Comment by mikedilger** [comment]
I am trying to replicate but upon following it is loading his relay list from wss://nostr21.com, so for the moment I'm commenting out subscribe_discover to get into the state he is in, then I will put that code back and see why it doesn't (after following is triggered) refresh the relay list.

**Comment by mikedilger** [comment]
Interesting.  The person record in the database has:
```
  "relay_list_created_at":null,
  "relay_list_last_received":1711401962
```

There may be an error/bug in relay list processing.

Here is his relay list:

```
{
  "id":"1ea4425ae1a641484d9b3570455c18397c73e56a66b85778186b13d3b56484f0",
  "pubkey":"7cc328a08ddb2afdf9f9be77beff4c83489ff979721827d628a542f32a247c0e",
  "created_at":1711131236,
  "kind":10002,
  "sig":"90971ebc348d2ea45d8584a605dba13009da7e4960b2fd9ad881370a2fd347ec4f2d04da4f92b1b30dd27a03e232ed95bc3dfa1607c2844bb8e7b3e76f8e9991",
  "content":"",
  "tags":[
    ["r","wss://bevo.nostr1.com/"],
    ["r","wss://hotrightnow.nostr1.com/"],
    ["r","wss://nos.lol/"],
    ["r","wss://nostr.thesamecat.io/"],
    ["r","wss://nostr.wine/"],
    ["r","wss://nostr21.com/"],
    ["r","wss://relay.damus.io/"],
    ["r","wss://theforest.nostr1.com/"]
  ]
}
```

**Comment by mikedilger** [comment]
Nope, that's a big mistake.  This field was repurposed very recently to be 'relay_list_last_sought', not when it was received.

So all it shows is that subscribe_discover() was run, it was attempted.  The purpose of this code change was so that it doesn't keep trying over and over to get a relay list that isn't there.  It needs to wait a while before it tries again, so it uses this field now to remember when it last tried, not when it last succeeded.  But I didn't rename the database field because that would be a schema update, I just put a serde rename tag on it (easier to code).

**Comment by jeremyd** [comment]
if there's a way to run some lmdb splunking, im all for it...

**Comment by mikedilger** [comment]
There are multiple command-line gossip commands that do most of the "lmdb splunking". Try running "gossip help".

**Comment by mikedilger** [comment]
Ok I found out what was happening in my case.  The relay wss://nostr21.com/ was not marked as a discovery relay, even though I chose it as such in the wizard.  When I marked it as discovery, the relay list showed up right away.

This is a wizard bug.

**Comment by mikedilger** [comment]
Huh, I can't replicate it now. Wizard works.

**Comment by jeremyd** [comment]
cool, didn't know about these commands, useful..

so for the user where it still says relay list not found, the print person shows

```
relay_list_created_at":1711304124,"relay_list_last_received":1711402501
```

**Comment by jeremyd** [comment]
what does print_person_relays, for a particular pubkey show?  for this pubkey, it's way more than the relay list for that pubkey is supposed to be.. it's basically all the relays that have been configured in this gossip client..  shouldn't it just be the relay list that was loaded from the followed pubkey's fetched relaylist?

**Comment by jeremyd** [comment]
there is a distinct difference for print_person_relays between the new test I did (where it found the list right away) and the other account im debuging.  the one that's broken, print_person_relays shows read: false, write: false for everything in the list, even for nostr21, and the list has hardly any relays that were from the actual list..

```
{
  "pubkey": "7cc328a08ddb2afdf9f9be77beff4c83489ff979721827d628a542f32a247c0e",
  "url": "wss://nostr21.com/",
  "last_fetched": 1711407742,
  "last_suggested_kind3": null,
  "last_suggested_nip05": 1711305047,
  "last_suggested_bytag": null,
  "read": false,
  "write": false,
  "manually_paired_read": false,
  "manually_paired_write": false
}
```

**Comment by jeremyd** [comment]
I guess, another thing I did, was add a LOT of relays, before adding some npubs.  I added the relays as Read and Discover.  Perhaps the bug can be exposed by adding more Read and Discover relays..

**Comment by mikedilger** [comment]
> cool, didn't know about these commands, useful..
> 
> so for the user where it still says relay list not found, the print person shows
> 
> ```
> relay_list_created_at":1711304124,"relay_list_last_received":1711402501
> ```

The "Relay list not found" is going off of the relays in the person_relay records. If none of them are marked as 'read' or 'write' then it either found an empty relay list or didn't properly process the one it found.

This code is changing. It won't be saying this anymore unless the relay list actually wasn't found.  But then when a relay list fails to process it will unfortunately think everything is fine.



> what does print_person_relays, for a particular pubkey show? for this pubkey, it's way more than the relay list for that pubkey is supposed to be.. it's basically all the relays that have been configured in this gossip client.. shouldn't it just be the relay list that was loaded from the followed pubkey's fetched relaylist?

If gossip somehow thought they were associated, it made a record. You can find out why gossip thought they were associated by looking at the fields and seeing which date stamp field has a value.

**Comment by mikedilger** [comment]
Oh, also if it fails to find relays for a person, then it falls back to your read relays.  Why not try reading from there if you have absolutely nothing else to go off of.

**Comment by bu5hm4nn** [comment]
> Why not try reading from there

Since we now are at the point that the overlord/minions can handle changing out the relay sets, we could also fall back to searching on all the recommended relays since we already have that list...

**Comment by jeremyd** [comment]
mentioned this on nostr, but not sure if you saw it.  So i updated to unstable branch.  And the notice went away on the follow list (there was no button).  When i manually checked print_person_relays, the list was unchanged from it's broken list (all relays were read: false, write: false still).

Not super urgent, but I'd like to be able to get gossip to refresh these though (kindof like in the past when you could force it to reload all metadata).  I already sent out a blast to everyone, and it may have reached them, or may not.  I will not be sending another for a few weeks so by then I'll either figure out how to fix my db, or re-create it.

**Comment by mikedilger** [comment]
I understand. This issue is still open. I've been pulled into other things but I will get back to it.

**Comment by mikedilger** [comment]
Ok I'm ready to look into this again and I can't remember where we left off.

**Comment by jeremyd** [comment]
Hey Mike,
Sorry was out of town for a few weeks.  This issue could just be a hiccup in my upgrade sequence or transient error when connecting to relays the first time..  I have since accidentally wiped both my gossip databases and was unable to reproduce on a new database.  From what I could tell, the data in print_person_relays just got corrupted but there was no way to make it re-crawl the data for them.  I think we left off thinking about how to possibly get it to refresh these should this happen again (with a refresh button or similar method).

**Comment by mikedilger** [comment]
Ok. Let's close this and next time it happens to somebody I'll dig in again.
